src/day10.scala

```diff
@@ -0,0 +1,118 @@
+// using scala 3.1.0
+
+package day10
+
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day10"))(_.mkString)
+
+enum CheckResult:
+  case Ok
+  case IllegalClosing(expected: Option[Symbol], found: Symbol)
+  case Incomplete(pending: List[Symbol])
+
+extension (illegalClosing: CheckResult.IllegalClosing)
+  def score: Int = 
+    import Symbol.*
+    illegalClosing.found match
+      case Parenthesis(_) => 3
+      case Bracket(_) => 57
+      case Brace(_) => 1197
+      case Diamond(_) => 25137
+
+enum SymbolState:
+  case Open, Close
+
+enum Symbol(val state: SymbolState):
+  case Parenthesis(override val state: SymbolState) extends Symbol(state)
+  case Bracket(override val state: SymbolState) extends Symbol(state)
+  case Brace(override val state: SymbolState) extends Symbol(state)
+  case Diamond(override val state: SymbolState) extends Symbol(state)
+
+  def isOpen: Boolean = state == SymbolState.Open
+
+  def opens(that: Symbol): Boolean =
+    import SymbolState.*
+    (this, that) match
+      case (Parenthesis(Open), Parenthesis(Close)) => true
+      case (Bracket(Open), Bracket(Close)) => true
+      case (Brace(Open), Brace(Close)) => true
+      case (Diamond(Open), Diamond(Close)) => true
+      case _ => false
```

Comment:
I wonder if the kind of symbol (`Parenthesis`, `Bracket`, etc.) should be a data of the model instead of being a “constructor”. Then `opens` would just be like so:

```suggestion
  def opens(that: Symbol): Boolean = this.kind == that.kind
```

Reply:
Basically, you would model an opening/closing symbol like so:

~~~ scala
enum Direction:
  case Open, Close
enum Kind:
  case Parenthesis, Bracket, Brace, Diamond
case class Symbol(kind: Kind, direction: Direction)
~~~

Reply:
I like this encoding. Simpler is better